contracts/PooledStaking.sol

```diff
@@ -0,0 +1,512 @@
+    // i.e. when pushing the first burn ever
+    // if those are different, burnCount - 1 will be >= 0 (will not underflow)
+    // but we should check that it doesn't point to an empty slot (processed & deleted burn)
+    if (burnCount != firstBurn && burns[burnCount - 1].burnedAt > 0) {
+      burns[burnCount - 1].next = burnCount; // set previous burn's next to current burn id
+    } else {
+      // otherwise this is the only existing burn, therefore it is the first one as well
+      firstBurn = burnCount;
+    }
+
+    ++burnCount; // update counter
+  }
+
+  function pushReward(address contractAddress, uint amount, address from) onlyInternal external {
+
+    // transfer tokens from specified contract to us
+    // token transfer should be approved by the specified address
+    Vault.deposit(token, from, amount);
+
+    rewards[rewardCount] = Reward(amount, now, contractAddress, 0); // add new reward
+
+    // TODO: recheck this
+    // similar to pushBurn
+    if (rewardCount != firstReward && rewards[rewardCount - 1].rewardedAt > 0) {
+      rewards[rewardCount - 1].next = rewardCount; // set previous reward's next to current id
+    } else {
+      firstReward = rewardCount;
+    }
+
+    ++rewardCount; // update counter
+  }
+
+  function processPendingActions() public {
+    while (true) {
+      Burn storage burn = burns[firstBurn];
+      Deallocation storage deallocation = deallocations[firstDeallocation];
+
+      bool canBurn = burn.burnedAt != 0; // end not reached
+      bool canDeallocate = deallocation.deallocateAt != 0 && deallocation.deallocateAt <= now;
+
+      if (!canBurn && !canDeallocate) {
+        // everything is processed
+        break;
+      }
+
+      // TODO: check if there's enough gas for the next iteration
+
+      // no burn to compare to or deallocation should happen before burn
+      if (!canBurn || deallocation.deallocateAt <= burn.burnedAt) {
+        _processFirstDeallocation();
+        continue;
+      }
+
+      _processFirstBurn();
+    }
+  }
+
+  function _processFirstDeallocation() internal {
+    Deallocation storage deallocation = deallocations[firstDeallocation];
+    Staker storage staker = stakers[deallocation.stakerAddress];
+
+    address contractAddress = deallocation.contractAddress;
+    staker.allocations[contractAddress].sub(deallocation.amount);
+    staker.pendingDeallocations[contractAddress].add(deallocation.amount);
+
+    uint nextDeallocation = deallocation.next;
+    delete deallocations[firstDeallocation];
+    firstDeallocation = nextDeallocation;
+  }
+
+  function _processFirstBurn() internal {
+
+    Burn storage burn = burns[firstBurn];
+    Contract storage _contract = contracts[burn.contractAddress];
+    address contractAddress = burn.contractAddress;
+
+    uint burnAmount = burn.amount <= _contract.staked ? burn.amount : _contract.staked;
+    uint burned = 0;
+    uint newContractStake = 0;
+
+    for (uint i = 0; i < _contract.stakers.length; i++) {
```

Comment:
is there any known upper bound to the _contract.stakers.length as of now?

we established for staker.contracts there is one that results from 

MIN_STAKE_PERCENTAGE and MAX_LEVERAGE

Reply:
worst case scenario, that's all members - that's around 700 atm, but I expect it to grow. now, this WC scenario is quite unlikely, as it's not economically viable at this point (i.e. small rewards distributed to too many people -> negligible rewards)

Reply:
was thinking in case we need to cap this for  the pool to perform properly we could agree on a feature level to a cap where only a "select club" can enter the pool based on some rules

Reply:
or first come first served and a limit on number of stakers per a particular contract